src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+---
+layout: default
+title: Workflow Non-deterministic errors
+permalink: /docs/go-client/workflow-non-deterministic-errors
+---
+
+# Workflow Non-deterministic errors
+
+## Root cause of non-deterministic errors
+Cadence workflows are designed as long-running operations, and therefore the workflow code you write must be deterministic so that no matter how many time it is executed it alwasy produce the same results.
+
+In production environment, your workflow code will run on a distributed system orchestrated by clusters of machines. However, machine failures are inevitable and can happen anytime to your workflow host. If you have a workflow running for long period of time, maybe months even years, and it fails due to loss of a host, it will be resumed on another machine and continue the rest of its execution.
+
+Consider the following diagram where `Workflow A` is running on `Host A` but suddenly it crashes.
+
+![change-workflow-ownership](../../shared/img/non-determinsitic-errors-page-visuals/change-workflow-ownership.png)
+
+Workflow A then will be picked up by Host B and continues its execution. This process is called <b>change of workflow ownership</b>. However, after Host B gains ownership of the Workflow A, it does not have any information about its historical executions. For example, Workflow A may have executed many activities and it fails. Host B needs to redo all its history until the moment of failure. The process of reconstructing history of a workflow is called <b>history replay</b>.
```

Comment:
>This process is called <b>change of workflow ownership</b>.

hmm.  do we call it that literally anywhere else?  it's also not really "ownership", it can move around randomly constantly and that's also normal and fine.